### PR TITLE
Make Partition ID non-mandatory during partition creation. 

### LINKED ISSUE #5014: Allow name-only partition creation
As a user of high amounts of tenancy, I have no interest in determining an actual partition ID, and would rather have the system auto-select an ID for me, given a name. Since all my logic is related to the tenant name, needing to know about the ID (and also select it) is a burden. It would be great if partition creation could make the ID optional, and auto-generated by the server.

### PR DESCRIPTION
**Pull request by tadgh**
closes #5014 

- Autogenerate a partition ID during create operation, if one is absent. 
- Changelogs
- Docs
- Test

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/partition/PartitionLookupSvcImpl.java [hunk 1]
```diff
@@ -120,9 +121,23 @@ public void clearCaches() {
...
 	}
 
+	/**
+	 * Generate a random int which is guaranteed to  be unused by an existing partition.
+	 * @return an integer representing a partition ID that is not currently in use by the system.
+	 */
+	public int generateRandomUnusedPartitionId() {
+		int candidate = ThreadLocalRandom.current().nextInt();
+		Optional<PartitionEntity> partition = myPartitionDao.findById(candidate);
+		while (partition.isPresent()) {
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/partition/PartitionLookupSvcImpl.java:131]: Does it really have the potential to perform many queries? The integer space is 2,147,483,647, even if we are creating 100,000 partitions the odds of a collision here (meaning you will need to do more than one query, is only 1−𝑘!𝑘𝑛(𝑘−𝑛)! This is the primary key of the field meaning lookup is pretty well instant. Also even if you create 15000 tenants, this is not a common operation.

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/partition/PartitionManagementProviderTest.java [hunk 2]
```diff
@@ -94,6 +98,34 @@ public void testCreatePartition() {
 		assertEquals("a description", ((StringType) response.getParameterValue(ProviderConstants.PARTITION_MANAGEMENT_PARTITION_DESC)).getValue());
 	}
 
+	@Test
+	public void testCreatePartitionWithNameOnlyAutogeneratesId() {
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/partition/PartitionManagementProviderTest.java:102]: Another PR is already doing this, I don't want to duplicate effort: https://github.com/hapifhir/hapi-fhir/issues/5011

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/docs/server_jpa_partitioning/partitioning_management_operations.md [hunk 3]
```diff
@@ -21,9 +21,9 @@ The `$partition-management-create-partition` operation can be used to create a n
         <tr>
             <td>id</td>
             <td>Integer</td>
-            <td>1..1</td>
+            <td>0..1</td>
             <td>
-                The numeric ID for the partition. This value can be any integer, positive or negative or zero. It must not be a value that has already been used. 
+                The numeric ID for the partition. This value can be any integer, positive or negative or zero. It must not be a value that has already been used. If omitted, a random integer will be selected.
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/docs/server_jpa_partitioning/partitioning_management_operations.md:26]: Good call, fixing

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/partition/PartitionLookupSvcImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/partition/PartitionLookupSvcImpl.java:131]: Well, I suppose it depends of how ThreadLocal Random will behave, but yes, I accept the probability shod be low. Feel free to leave it like this.

**Review by jmarchionatto — APPROVED** [review]

**Pull request merged**